topdown/resolver.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2020 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package topdown
+
+import (
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/metrics"
+	"github.com/open-policy-agent/opa/resolver"
+)
+
+type resolverTrie struct {
+	r        resolver.Resolver
+	children map[ast.Value]*resolverTrie
+}
+
+func newResolverTrie() *resolverTrie {
+	return &resolverTrie{children: map[ast.Value]*resolverTrie{}}
+}
+
+func (t *resolverTrie) Put(ref ast.Ref, r resolver.Resolver) {
+	node := t
+	for _, t := range ref {
+		child, ok := node.children[t.Value]
+		if !ok {
+			child = &resolverTrie{children: map[ast.Value]*resolverTrie{}}
+			node.children[t.Value] = child
+		}
+		node = child
+	}
+	node.r = r
+}
+
+func (t *resolverTrie) Resolve(e *eval, ref ast.Ref) (ast.Value, error) {
+	e.metrics.Timer(metrics.RegoExternalResolve).Start()
+	defer e.metrics.Timer(metrics.RegoExternalResolve).Stop()
+
+	in := resolver.Input{
+		Ref:     ref,
```

Comment:
This needs to change so that we only pass the entrypoint ref--not the entire ref. We can set the ref below:

```
in.Ref = ref[:i+1]
result, err := node.r.Eval(e.ctx, in)
```

Reply:
Sent a PR against the fork to fix this: https://github.com/patrick-east/opa/pull/21